src/Microsoft.AspNetCore.Mvc.TagHelpers/EnvironmentTagHelper.cs

```diff
@@ -68,27 +92,75 @@ public override void Process(TagHelperContext context, TagHelperOutput output)
+                    {
+                        hasExcludeEnvironments = true;
+                        if (environment.Equals(currentEnvironmentName))
+                        {
+                            output.SuppressOutput();
+                            return;
+                        }
+                    }
+                }
+            }
+
             var hasEnvironments = false;
-            foreach (var item in tokenizer)
+            if (Names != null)
             {
-                var environment = item.Trim();
-                if (environment.HasValue && environment.Length > 0)
+                var tokenizer = new StringTokenizer(Names, NameSeparator);
+                foreach (var item in tokenizer)
                 {
-                    hasEnvironments = true;
-                    if (environment.Equals(currentEnvironmentName, StringComparison.OrdinalIgnoreCase))
+                    var environment = item.Trim();
+                    if (environment.HasValue && environment.Length > 0)
                     {
-                        // Matching environment name found, do nothing
-                        return;
+                        hasEnvironments = true;
+                        if (environment.Equals(currentEnvironmentName, StringComparison.OrdinalIgnoreCase))
+                        {
+                            // Matching environment name found, do nothing
+                            return;
+                        }
                     }
                 }
             }
 
-            if (hasEnvironments)
+            var hasIncludeEnvironments = false;
+            if (Include != null)
             {
-                // This instance had at least one non-empty environment specified but none of these
+                var tokenizer = new StringTokenizer(Include, IncludeEnvironmentsSeparator);
+                foreach (var item in tokenizer)
+                {
+                    var environment = item.Trim();
+                    if (environment.HasValue && environment.Length > 0)
+                    {
+                        hasIncludeEnvironments = true;
+                        if (environment.Equals(currentEnvironmentName))
+                        {
+                            // Matching environment name found, do nothing
+                            return;
+                        }
+                    }
+                }
+            }
+
+            if (hasEnvironments || hasIncludeEnvironments)
+            {
+                // This instance had at least one non-empty environment (name or include) specified but none of these
                 // environments matched the current environment. Suppress the output in this case.
                 output.SuppressOutput();
+                return;
+            }
+
+            if (hasExcludeEnvironments)
+            {
+                // No matching exclude environment name found, do nothing
+                return;
             }
```

Comment:
I partially agree with this comment. I think the return at line 157 is important because it ensures the method will end right there and then. It will prevent a future change later in this method from inadvertently running.